rust/main/utils/run-locally/src/sealevel/mod.rs

```diff
@@ -0,0 +1,328 @@
+        )
+        .join();
+        state.data.push(Box::new(solana_path_tempdir));
+
+        let solana_program_builder = build_solana_programs(solana_bin_path.clone());
+        (solana_bin_path, solana_program_builder.join())
+    };
+
+    // this task takes a long time in the CI so run it in parallel
+    log!("Building rust...");
+    let build_main = Program::new("cargo")
+        .cmd("build")
+        .working_dir(&workspace_path)
+        .arg("features", "test-utils")
+        .arg("bin", "relayer")
+        .arg("bin", "validator")
+        .arg("bin", "scraper")
+        .arg("bin", "init-db")
+        .filter_logs(|l| !l.contains("workspace-inheritance"))
+        .run();
+
+    log!("Running postgres db...");
+    let postgres = Program::new("docker")
+        .cmd("run")
+        .flag("rm")
+        .arg("name", "scraper-testnet-postgres")
+        .arg("env", "POSTGRES_PASSWORD=47221c18c610")
+        .arg("publish", "5432:5432")
+        .cmd("postgres:14")
+        .spawn("SQL", None);
+    state.push_agent(postgres);
+
+    build_main.join();
+
+    log!("Building hyperlane-sealevel-client...");
+    Program::new("cargo")
+        .working_dir(&sealevel_path)
+        .cmd("build")
+        .arg("bin", "hyperlane-sealevel-client")
+        .filter_logs(|l| !l.contains("workspace-inheritance"))
+        .run()
+        .join();
+
+    let solana_ledger_dir = tempdir().expect("Failed to create solana ledger dir");
+    let (solana_cli_tools_path, solana_config_path) = {
+        // use the agave 2.x validator version to ensure mainnet compatibility
+        let solana_tools_dir = tempdir().expect("Failed to create solana tools dir");
+        let solana_bin_path = install_solana_cli_tools(
+            SOLANA_NETWORK_CLI_RELEASE_URL.to_owned(),
+            SOLANA_NETWORK_CLI_VERSION.to_owned(),
+            solana_tools_dir.path().to_path_buf(),
+        )
+        .join();
+        state.data.push(Box::new(solana_tools_dir));
+
+        let start_solana_validator = start_solana_test_validator(
+            solana_bin_path.clone(),
+            hyperlane_solana_programs_path.clone(),
+            solana_ledger_dir.as_ref().to_path_buf(),
+        );
+
+        let (solana_config_path, solana_validator) = start_solana_validator.join();
+        state.push_agent(solana_validator);
+        (solana_bin_path, solana_config_path)
+    };
+
+    sleep(Duration::from_secs(5));
+
+    log!("Init postgres db...");
+    Program::new(concat_path(&workspace_path, "target/debug/init-db"))
+        .working_dir(&workspace_path)
+        .run()
+        .join();
+    state.push_agent(scraper_env.spawn("SCR", None));
+
+    // spawn the rest of the validators
+    for (i, validator_env) in validator_envs.into_iter().enumerate() {
+        let validator = validator_env.spawn(
+            make_static(format!("VL{}", 1 + i)),
+            Some(AGENT_LOGGING_DIR.as_ref()),
+        );
+        state.push_agent(validator);
+    }
```

Comment:
would be good to spin up the validators after the first half of messages are sent, to test both the forward and the backward indexing. So right before the relayer is started, 10 lines below